cpp/src/arrow/record_batch.cc

```diff
@@ -247,6 +248,109 @@ Result<std::shared_ptr<StructArray>> RecordBatch::ToStructArray() const {
                                        /*offset=*/0);
 }
 
+template <typename DataType>
+inline void ConvertColumnsToTensor(const RecordBatch& batch, uint8_t* out) {
+  using CType = typename arrow::TypeTraits<DataType>::CType;
+  auto* out_values = reinterpret_cast<CType*>(out);
+
+  // Loop through all of the columns
+  for (int i = 0; i < batch.num_columns(); ++i) {
+    const auto& arr = *batch.column(i);
+    auto data = arr.data();
+    const auto data_offset = data->offset * sizeof(CType);
+    const auto& in_values =
+        reinterpret_cast<const CType*>(data->buffers[1]->data() + data_offset);
```

Comment:
Is this the same as `data->GetValues<CType>(1)` ?

Reply:
Seems so. If I change the line 262 and remove the definition of `data_offset` all tests pass.